code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,394 @@
+				boutput(user, "<span class='alert'>The [W.name] doesn't contain any oxygen.</span>")
+				return
+
+			src.visible_message("<span class='notice'>[user] loads [W] into the [src].</span>")
+			user.u_equip(W)
+			W.set_loc(src)
+			src.inlet_tank = W
+			src.UpdateIcon()
+			playsound(src.loc, "sound/items/Deconstruct.ogg", 50, 1)
+
+
+		// fuel tank
+		else if (istype(W, /obj/item/reagent_containers/food/drinks/fueltank))
+			if (src.fuel_tank)
+				boutput(user, "<span class='alert'>There appears to be a fuel tank loaded already!</span>")
+				return
+
+			if (!src.get_average_volatility(W.reagents))
+				boutput(user, "<span class='alert'>The [W.name] doesn't contain any fuel!</span>")
+				return
+
+			src.visible_message("<span class='notice'>[user] loads [W] into the [src].</span>")
+			user.u_equip(W)
+			W.set_loc(src)
+			src.fuel_tank = W
+			src.UpdateIcon()
+			playsound(src.loc, "sound/items/Deconstruct.ogg", 50, 1)
+
+		else if (iswrenchingtool(W))
+			if (src.anchored)
+				if (src.active)
+					src.visible_message("<span class='notice'>The [src] stops as it was unachored by [user].</span>")
+					src.stop_engine()
+				else
+					src.visible_message("<span class='notice'>[user] secures the [src]'s bolts into the floor.</span>")
+
+				src.anchored = 0
+				playsound(src.loc, "sound/items/Ratchet.ogg", 50, 1)
+				return
+
+			src.anchored = 1
+			src.visible_message("<span class='notice'>[user] removes the [src]'s bolts from the floor.</span>")
+			playsound(src.loc, "sound/items/Ratchet.ogg", 50, 1)
+
+		src.updateUsrDialog()
+		return
+
+#define FUEL_QUALITY_VERY_HIGH 10
+#define FUEL_QUALITY_HIGH 6
+#define FUEL_QUALITY_AVERAGE 3
+#define FUEL_QUALITY_LOW 2
+#define FUEL_QUALITY_VERY_LOW 1
+
+	process()
+		if (!src || !src.active)
+			return
+
+		var/average_volatility = get_average_volatility(src.fuel_tank.reagents)
+		var/available_oxygen = src.check_available_oxygen()
+		if (!available_oxygen || !average_volatility)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src]'s engine fails to run, it has nothing to combust!</span>")
+			return
+
+		if (!src.anchored)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src] makes a horrible racket and shuts down, it has become unanchored!</span>")
+			return
+
+		var/obj/cable/C = src.get_output_cable()
+		if (!C)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>Electricity begins to arc off the [src] causing it to shutdown, it has nothing to output to!</span>")
+			elecflash(src.loc, 0, power = 3, exclude_center = 0)
+			return
+
+		src.last_output = ((average_volatility * (available_oxygen * 5)) * src.output_multiplier) KILO WATTS
+		var/datum/powernet/P = C.get_powernet()
+		P.newavail += src.last_output
+
+		var/turf/simulated/T = get_turf(src)
+		if (istype(T))
+			var/datum/gas_mixture/payload = new /datum/gas_mixture
+			payload.carbon_dioxide = src.atmos_drain_rate * average_volatility * src.output_multiplier
```

Comment:
Given how bad CO2 is to deal with, this is a huge downside that would basically just require you to always run an air filter next to the generator

Reply:
The amount that it outputs is very small, and takes quite some time to buildup. There is also a warning in the description. I feel it wouldn't make much sense for their to be no byproduct from burning the reagents.

Couple ideas would be to replace it with dumping some ash into the fueltank or adding a warning when the concentration of CO2 gets too high, like some alarms or flashing lights.